zeus/monitor.py

```diff
@@ -0,0 +1,169 @@
+        zeus_monitor_sleep_ms: int = 0,
+        zeus_monitor_log_dir: str | None = None,
+    ) -> None:
+        """Create a Zeus monitor context.
+
+        Args:
+            skip_steps: The number of steps to skip when profiling energy consumption.
+            profile_steps: The number of steps to profile and average over for energy consumption.
+            device_id: CUDA device ID to run the monitor for.
+            zeus_monitor_path: `argv[0]` to use when spawning the Zeus monitor.
+            zeus_monitor_sleep_ms: How long the Zeus monitor should sleep after sampling power.
+            zeus_monitor_log_dir: The directory to put the monitor log file. A temporary file is
+                used if not specified.
+        """
+        # Save arguments.
+        self.skip_steps = skip_steps
+        self.profile_steps = profile_steps
+        self.device_id = device_id
+
+        # Spawn the Zeus monitor.
+        if zeus_monitor_log_dir:
+            self._power_csv = f"{zeus_monitor_log_dir}/{self.device_id}.power.csv"
+        else:
+            self._power_csv = tempfile.mkstemp(
+                suffix=f"+gpu{self.device_id}.power.csv"
+            )[1]
+        self._monitor = subprocess.Popen(
+            args=[
+                zeus_monitor_path,
+                self._power_csv,
+                "0",
+                str(zeus_monitor_sleep_ms),
+                str(self.device_id),
+            ],
+            stdin=subprocess.DEVNULL,
+        )
+        self._time_origin = time.monotonic()
+
+        # Make sure the monitor is eventually stopped.
+        def exit_hook():
+            self._monitor.send_signal(signal.SIGINT)
+            time.sleep(2.0)
+            self._monitor.kill()
+
+        atexit.register(exit_hook)
+
+        # Set internal profiling states.
+        self._current_step = 0
+        self._profile_start_time = 0.0
+        self._profile_end_time = 0.0
+
+    def start_step(self) -> None:
+        """Mark the beginning of one step."""
+        current_time = time.monotonic()
+        self._current_step += 1
+        if self._current_step == self.skip_steps + 1:
+            self._profile_start_time = current_time - self._time_origin
+
+    def finish_step(self) -> None:
+        """Mark the end of one step."""
+        current_time = time.monotonic()
+        if self._current_step == self.skip_steps + self.profile_steps:
+            self._profile_end_time = current_time - self._time_origin
+
+    @contextmanager
+    def step(self) -> Generator[None, None, None]:
+        """Wrap one training step to mark start and finish times."""
+        try:
+            self.start_step()
+            yield
+        finally:
+            self.finish_step()
+
+    @property
+    def is_done(self) -> bool:
+        """Return whether the specified profiling steps are done."""
+        return self._current_step >= self.skip_steps + self.profile_steps
+
+    @property
+    def energy(self) -> float:
```

Comment:
Do we need to prevent users from querying energy/time if the profiling is not done? I just recall that if the period between `start` and `end` is too short, there can be an exception (basically, no energy logging between `start` and `end`).

Reply:
Yeah, that actually makes things a bit more complicated. So, because data loading happens between two iterations, we need to accumulate the *training* time between each pair of `start_step` and `end_step`, and accumulate the energy consumed by *training* (energy located at the last recorded time window) at `end_step`, right?
Also, do we need some mechanism to force the ordering of `start_step` -> `end_step`? Maybe an additional bool?

P.S. Nice catch on the `cuda.synchronize()`!!

Reply:
Yeah so basically we should do `self.profile_steps` number of calls to `compute_energy`, with the start and finish time of each step. Regarding the ordering of `start_step` and `finish_step`, honestly I think we're good without it. It's difficult to imagine people calling `start_step` two times 😁